### PR TITLE
bullet-featherstone: fix how links are flattened in ConstructSdfModel

### PR DESCRIPTION
**Pull request by shameekganguly**
Fixes a segfault in `ConstructSdfModel` when loading a model with links defined out of order. Here's an example model that triggers the segfault:
```
  <model name="test_model">
      <link name="child_linkC" />
      <link name="parent_link" />
      <link name="child_linkA" />
      <link name="child_linkB" />
      <joint name="world_joint" type="fixed">
        <parent>world</parent>
        <child>parent_link</child>
      </joint>
      <joint name="joint1" type="prismatic">
        <parent>parent_link</parent>
        <child>child_linkA</child>
        <axis>
            <xyz>0 0 1</xyz>
        </axis>
      </joint>
      <joint name="joint2" type="prismatic">
        <parent>child_linkA</parent>
        <child>child_linkB</child>
        <axis>
            <xyz>0 0 1</xyz>
        </axis>
      </joint>
      <joint name="joint3" type="prismatic">
        <parent>child_linkB</parent>
        <child>child_linkC</child>
        <axis>
            <xyz>0 0 1</xyz>
        </axis>
      </joint>
    </model>
```

The current logic for sorting links to ensure that a parent link is placed in the `flatLinks` vector before its children links is inadequate for such cases. So the fix is to build the link tree and do a preorder traversal of it to create the `flatLinks` vector.

### CODE REVIEW

**Review by scpeters — COMMENTED** [review]
we should add a test that uses the snippet you provided. I'll look into the easiest way to do that

**Comment by scpeters** [comment]
> we should add a test that uses the snippet you provided. I'll look into the easiest way to do that

I added a test in https://github.com/gazebosim/gz-physics/commit/49de232c10afa76541b03221bd14778cfb29a9d6 without this branch, and it fails CI. I will push it to this branch

**Comment by scpeters** [comment]
@azeey I added a small test for this in case you want to review

File: bullet-featherstone/src/SDFFeatures.cc [hunk 1]
```diff
@@ -135,9 +135,11 @@ std::optional<Structure> ValidateModel(const ::sdf::Model &_sdfModel)
   const ::sdf::Joint *rootJoint = nullptr;
   bool fixed = false;
   const std::string &rootModelName = _sdfModel.Name();
+  std::unordered_map<const ::sdf::Link*, std::vector<const ::sdf::Link*>>
+    linkTree;
```
Comment (scpeters) [inline bullet-featherstone/src/SDFFeatures.cc:139]: A comment describing the contents of `linkTree` would be helpful; I believe it is a map from each parent link to a vector of its child links.

File: bullet-featherstone/src/SDFFeatures.cc [hunk 2]
```diff
@@ -254,23 +260,30 @@ std::optional<Structure> ValidateModel(const ::sdf::Model &_sdfModel)
...
-        linkIndex[link] = linkIndex.size();
-        flatLinks.push_back(link);
       }
 
       return true;
     };
 
-  if (!flattenLinks(_sdfModel))
+  if (rootLink == nullptr && !findRootLink(_sdfModel)) {
     return std::nullopt;
```
Comment (scpeters) [inline bullet-featherstone/src/SDFFeatures.cc:270]: add comment: "No root link was found in this model"

File: bullet-featherstone/src/SDFFeatures.cc [hunk 3]
```diff
@@ -254,23 +260,30 @@ std::optional<Structure> ValidateModel(const ::sdf::Model &_sdfModel)
...
 
   for (std::size_t i = 0; i < _sdfModel.ModelCount(); ++i)
   {
-    if (!flattenLinks(*_sdfModel.ModelByIndex(i)))
+    if (rootLink != nullptr) {
+      break;
+    }
+    if (!findRootLink(*_sdfModel.ModelByIndex(i))) {
       return std::nullopt;
+    }
```
Comment (scpeters) [inline bullet-featherstone/src/SDFFeatures.cc:280]: can you add comments for this for loop? It looks like it searches for a root link in nested models if one is not already found

File: bullet-featherstone/src/SDFFeatures.cc (hunk 1 above)
Comment (iche033) [reply bullet-featherstone/src/SDFFeatures.cc:140]: added comment. [e689f8b](https://github.com/gazebosim/gz-physics/pull/562/commits/e689f8b9513c7aa3e94eea73b2979f883699cf3d)

File: bullet-featherstone/src/SDFFeatures.cc (hunk 2 above)
Comment (iche033) [reply bullet-featherstone/src/SDFFeatures.cc:279]: added. [e689f8b](https://github.com/gazebosim/gz-physics/pull/562/commits/e689f8b9513c7aa3e94eea73b2979f883699cf3d)

File: bullet-featherstone/src/SDFFeatures.cc (hunk 3 above)
Comment (iche033) [reply bullet-featherstone/src/SDFFeatures.cc:292]: added comment. [e689f8b](https://github.com/gazebosim/gz-physics/pull/562/commits/e689f8b9513c7aa3e94eea73b2979f883699cf3d)

**Review by scpeters — APPROVED** [review]
LGTM, though I think the test I added deserves a second look from someone else if possible

**Review by azeey — APPROVED** [review]
LGTM! Just a few minor comments.

File: bullet-featherstone/src/SDFFeatures.cc [hunk 4]
```diff
@@ -217,6 +220,17 @@ std::optional<Structure> ValidateModel(const ::sdf::Model &_sdfModel)
...
+        if (parent != nullptr)
+        {
+          if (linkTree.count(parent) == 0)
+          {
+            linkTree[parent] = {child};
+          }
+          else
+          {
+            linkTree[parent].push_back(child);
+          }
```
[the remark this answers is not in the document]
Comment (iche033) [reply bullet-featherstone/src/SDFFeatures.cc]: updated. [ef0ef82](https://github.com/gazebosim/gz-physics/pull/562/commits/ef0ef821a75253f96b54eb1375e3f67eb1b9692b)

File: test/common_test/world_features.cc [hunk 5]
```diff
@@ -182,19 +179,59 @@ TYPED_TEST(WorldFeaturesTest, GravityFeatures)
...
+      gz::common::joinPaths(TEST_WORLD_DIR, "world_unsorted_links.world"));
+    EXPECT_TRUE(errors.empty()) << errors;
+    const sdf::World *sdfWorld = root.WorldByIndex(0);
+    EXPECT_NE(nullptr, sdfWorld);
+
+    // Per https://github.com/gazebosim/gz-physics/pull/562, there is a
+    // seg-fault in the bullet-featherstone implementation of ConstructSdfModel
+    // (called by ConstructSdfWorld). So loading the world successfully
+    // is enough for this test.
+    auto world = engine->ConstructWorld(*root.WorldByIndex(0));
```
[the remark this answers is not in the document]
Comment (iche033) [reply test/common_test/world_features.cc]: updated. [ef0ef82](https://github.com/gazebosim/gz-physics/pull/562/commits/ef0ef821a75253f96b54eb1375e3f67eb1b9692b)

File: test/common_test/world_features.cc [hunk 6]
```diff
@@ -182,19 +179,59 @@ TYPED_TEST(WorldFeaturesTest, GravityFeatures)
...
+    ASSERT_NE(nullptr, engine);
+    EXPECT_TRUE(engine->GetName().find(this->PhysicsEngineName(name)) !=
+                std::string::npos);
+
+    sdf::Root root;
+    const sdf::Errors errors = root.Load(
+      gz::common::joinPaths(TEST_WORLD_DIR, "world_unsorted_links.world"));
+    EXPECT_TRUE(errors.empty()) << errors;
+    const sdf::World *sdfWorld = root.WorldByIndex(0);
+    EXPECT_NE(nullptr, sdfWorld);
```
[the remark this answers is not in the document]
Comment (iche033) [reply test/common_test/world_features.cc]: [ef0ef82](https://github.com/gazebosim/gz-physics/pull/562/commits/ef0ef821a75253f96b54eb1375e3f67eb1b9692b)

File: test/common_test/world_features.cc [hunk 7]
```diff
@@ -182,19 +179,59 @@ TYPED_TEST(WorldFeaturesTest, GravityFeatures)
...
-                               gz::physics::RemoveEntities,
-                               gz::physics::GetNestedModelFromModel,
-                               gz::physics::sdf::ConstructSdfLink,
-                               gz::physics::sdf::ConstructSdfJoint,
-                               gz::physics::sdf::ConstructSdfModel,
-                               gz::physics::sdf::ConstructSdfNestedModel,
-                               gz::physics::ConstructEmptyLinkFeature,
-                               gz::physics::ConstructEmptyNestedModelFeature
-                               >
+using ConstructModelFeatures = gz::physics::FeatureList<
```
[the remark this answers is not in the document]
Comment (iche033) [reply test/common_test/world_features.cc]: changed to `struct` [ef0ef82](https://github.com/gazebosim/gz-physics/pull/562/commits/ef0ef821a75253f96b54eb1375e3f67eb1b9692b)

File: test/common_test/world_features.cc [hunk 8]
```diff
@@ -182,19 +179,59 @@ TYPED_TEST(WorldFeaturesTest, GravityFeatures)
...
+    // Per https://github.com/gazebosim/gz-physics/pull/562, there is a
+    // seg-fault in the bullet-featherstone implementation of ConstructSdfModel
+    // (called by ConstructSdfWorld). So loading the world successfully
+    // is enough for this test.
+    auto world = engine->ConstructWorld(*root.WorldByIndex(0));
+    EXPECT_NE(nullptr, world);
+  }
+}
+
+using WorldModelFeatureList = gz::physics::FeatureList<
```
[the remark this answers is not in the document]
Comment (iche033) [reply test/common_test/world_features.cc]: changed to `struct`. [ef0ef82](https://github.com/gazebosim/gz-physics/pull/562/commits/ef0ef821a75253f96b54eb1375e3f67eb1b9692b)

File: test/common_test/worlds/world_unsorted_links.world
[the remark this answers is not in the document]
Comment (iche033) [reply test/common_test/worlds/world_unsorted_links.world:1]: renamed to .sdf extension. [ef0ef82](https://github.com/gazebosim/gz-physics/pull/562/commits/ef0ef821a75253f96b54eb1375e3f67eb1b9692b)

File: test/common_test/worlds/world_unsorted_links.world [hunk 9]
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" ?>
+<sdf version="1.6">
+  <world name="default">
+    <plugin
+      filename="gz-sim-physics-system"
+      name="gz::sim::systems::v0::Physics">
```
[the remark this answers is not in the document]
Comment (iche033) [reply test/common_test/worlds/world_unsorted_links.world]: removed. [ef0ef82](https://github.com/gazebosim/gz-physics/pull/562/commits/ef0ef821a75253f96b54eb1375e3f67eb1b9692b)

File: test/common_test/world_features.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (scpeters) [reply test/common_test/world_features.cc]: ok, makes sense. I had copied the approach from #493, so you may consider updating that to use the `struct` approach if you're planning to come back to it

**Comment by shameekganguly** [comment]
Thanks Steve, Ian, Addisu for taking care of this!

**Pull request merged**